sklearn/tests/test_impute.py

```diff
@@ -1030,6 +1033,593 @@ def test_missing_indicator_sparse_param(arr_type, missing_values,
+    assert_array_almost_equal(imputer.statistics_, statistics_mean)
+
+    # Test with varying missingness patterns
+    X = np.array([
+        [1,         0,          0,  1],
+        [0,         np.nan,     1,  np.nan],
+        [1,         1,          1,  np.nan],
+        [0,         1,          0,  0],
+        [0,         0,          0,  0],
+        [1,         0,          1,  1],
+        [10,        10,         10, 10],
+    ])
+    statistics_mean = np.nanmean(X, axis=0)
+
+    # Get weights of donor neighbors
+    dist = nan_euclidean_distances(X)
+    r1c1_nbor_dists = dist[1, [0, 2, 3, 4, 5]]
+    r1c3_nbor_dists = dist[1, [0, 3, 4, 5, 6]]
+    r1c1_nbor_wt = (1/r1c1_nbor_dists)
+    r1c3_nbor_wt = (1 / r1c3_nbor_dists)
+
+    r2c3_nbor_dists = dist[2, [0, 3, 4, 5, 6]]
+    r2c3_nbor_wt = 1/r2c3_nbor_dists
+
+    # Collect donor values
+    col1_donor_values = np.ma.masked_invalid(X[[0, 2, 3, 4, 5], 1]).copy()
+    col3_donor_values = np.ma.masked_invalid(X[[0, 3, 4, 5, 6], 3]).copy()
+
+    # Final imputed values
+    r1c1_imp = np.ma.average(col1_donor_values, weights=r1c1_nbor_wt)
+    r1c3_imp = np.ma.average(col3_donor_values, weights=r1c3_nbor_wt)
+    r2c3_imp = np.ma.average(col3_donor_values, weights=r2c3_nbor_wt)
+
+    X_imputed = np.array([
+        [1,         0,          0,  1],
+        [0,         r1c1_imp,   1,  r1c3_imp],
+        [1,         1,          1,  r2c3_imp],
+        [0,         1,          0,  0],
+        [0,         0,          0,  0],
+        [1,         0,          1,  1],
+        [10,        10,         10, 10],
+    ])
+
+    imputer = KNNImputer(weights="distance")
+    assert_array_almost_equal(imputer.fit_transform(X), X_imputed, decimal=6)
+    assert_array_almost_equal(imputer.statistics_, statistics_mean)
+
+
+def test_knn_imputer_metric_type():
+    X = np.array([
+        [0,      0],
+        [np.nan, 2],
+    ])
+
+    # Test with a metric type without NaN support
+    imputer = KNNImputer(metric="euclidean")
+    bad_metric_msg = "The selected metric does not support NaN values"
+    with pytest.raises(ValueError, match=bad_metric_msg):
+        imputer.fit(X)
+
+
+def test_knn_imputer_callable_metric():
+
+    # Define callable metric that returns the l1 norm:
+    def custom_callable(x, y, missing_values=np.nan, squared=False):
+        x = np.ma.array(x, mask=np.isnan(x))
+        y = np.ma.array(y, mask=np.isnan(y))
+        dist = np.nansum(np.abs(x-y))
+        return dist
+
+    X = np.array([
+        [4, 3, 3, np.nan],
+        [6, 9, 6, 9],
+        [4, 8, 6, 9],
+        [np.nan, 9, 11, 10.]
+    ])
+
+    X_imputed = np.array([
+        [4, 3, 3, 9],
+        [6, 9, 6, 9],
+        [4, 8, 6, 9],
+        [5, 9, 11, 10.]
+    ])
+
+    imputer = KNNImputer(n_neighbors=2, metric=custom_callable)
+    assert_array_almost_equal(imputer.fit_transform(X), X_imputed)
+
+
+def test_knn_imputer_complete_features():
+
+    # Test with use_complete=True
```

Comment:
What does this mean?

Reply:
This has been removed. (Must have been an old parameter)